### ISSUE TITLE
LoRA-GGPO

### ISSUE THREAD

**Issue by rockerBOO** [LoganBooker/prodigy-plus-schedule-free#17]
https://arxiv.org/abs/2502.14538v1

Code isn't available yet but they propose using the weight norm and gradient norm to add perturbations to the loss. So my question would be how to get the grad norm appropriately. Because there is also clipping by grad norm but noted to not use grad norm clipping as this optimizer handles it in some way. 

I'm not quite sure what I'm asking but if you're interested in it maybe some way of getting the gradient norm of A/B for LoRA training. 

For example I tried running it through Claude

```python
    def forward(self, x):
        # Base output with frozen weights
        base_output = torch.nn.functional.linear(x, self.weight)
        if self.bias is not None:
            base_output += self.bias.unsqueeze(0).expand_as(base_output)
        
        # LoRA component
        lora_output = (x @ self.lora_A) @ self.lora_B
        
        # During training, apply GGPO perturbation
        if self.training and hasattr(self, 'perturbation_seed'):
            # Generate perturbations using stored norms and seed
            torch.manual_seed(self.perturbation_seed)
            combined_weight = self.weight + self.scaling * (self.lora_B.T @ self.lora_A.T)
            perturbation_scale = torch.sqrt(self.weight_norms + 0.01 * self.grad_norms)
            perturbation = torch.randn_like(combined_weight) * perturbation_scale / math.sqrt(combined_weight.shape[0])
            
            # Apply perturbation to output
            perturbation_output = x @ perturbation.T
            return base_output + self.scaling * lora_output + perturbation_output
        else:
            # Regular LoRA forward pass
            return base_output + self.scaling * lora_output

  def update_norms(self, optimizer, sigma=0.03):
        # Extract combined weights
        combined_weight = self.weight + self.scaling * (self.lora_B.T @ self.lora_A.T)
        
        # Calculate weight norms
        self.weight_norms = torch.norm(combined_weight, dim=1, keepdim=True) ** 2
        
        # Extract gradients from optimizer
        # For LoRA A
        if self.lora_A.grad is not None:
            grad_A = self.lora_A.grad
            # For LoRA B
            grad_B = self.lora_B.grad
            
            # Approximate gradient for combined weight
            approx_grad = self.scaling * (grad_B.T @ self.lora_A.T + self.lora_B.T @ grad_A.T)
            
            # Calculate gradient norms
            self.grad_norms = torch.norm(approx_grad, dim=1, keepdim=True) ** 2
        
        # Set perturbation seed for reproducibility
        self.perturbation_seed = torch.randint(0, 2**32 - 1, (1,)).item()
```

Is getting the norm in this way appropriate with the clip grad norm from this optimizer?
